tests/smoke_tests/run_smoke_test.py

```diff
@@ -304,6 +342,70 @@ async def _wait_for_process_to_finish_and_retrieve_logs(
     return full_output
 
 
+class MetricType(Enum):
+    SERVER = "server"
+    CLIENT = "client"
+
+
+DEFAULT_METRICS_FOLDER = Path("metrics")
+DEFAULT_TOLERANCE = 0.0005
+
+
+def _assert_metrics(metric_type: MetricType, metrics_to_assert: Optional[Dict[str, Any]] = None) -> None:
+    if metrics_to_assert is None:
+        return
+
+    metrics_found = False
+    for file in DEFAULT_METRICS_FOLDER.iterdir():
+        if not file.is_file() or not str(file).endswith(".json"):
+            continue
+
+        with open(file) as f:
+            metrics = json.load(f)
+
+        if metrics["type"] != metric_type.value:
+            continue
+
+        metrics_found = True
+        _assert_metrics_dict(metrics_to_assert, metrics)
```

Comment:
I may be missing something, so correct me if I'm wrong, but this looks like we're going to compare any metrics json dumped by a client to a single `metrics_to_assert` dictionary. For some of our example, that's definitely fine, because the different clients are loading the same dataset and running the same training cycle. However, there are instances where we might want to test when that isn't the case. That is, each client loads a distinct dataset. So their metrics won't be the same as each other.

Again, correct me if that interpretation is wrong.

Reply:
Yes, I think that makes sense. We can do another smoke test where two clients load slightly different datasets and include it as a separate test.